### PR TITLE
Create ZIP archive from native image to allow distribution via SDKMAN!

### LINKED ISSUE #48: List mvnd on sdkman.io
Citing @gastaldi:

> it would be nice to have it listed in sdkman.io

Volunteers?

### PR DESCRIPTION
**Pull request by helpermethod**
Fixes #48.

### CODE REVIEW

**Comment by helpermethod** [comment]
I love it! 

If I am not totally mistaken `SDKMAN!` supports postinstall hooks, so this should be doable!

**Comment by helpermethod** [comment]
Hi @ppalaga,

here are some of my thoughts after having talked to the @marcOder:

There are several requirements when it comes to distribution via SDKMAN!

* there needs to be an enclosing folder, that could be something like `mvnd-dist-0.0.2-darwin-amd64`
* the binary needs to be present in the `bin` folder directly under the enclosing folder

Thus I propose the following ZIP structure

```
mvnd-0.0.2-darwin-amd64/
mvnd-0.0.2-darwin-amd64/bin
mvnd-0.0.2-darwin-amd64/bin/mvnd
mvnd-0.0.2-darwin-amd64/daemon/
mvnd-0.0.2-darwin-amd64/daemon/boot/
mvnd-0.0.2-darwin-amd64/daemon/lib/
mvnd-0.0.2-darwin-amd64/daemon/bin
...
```

**Comment by ppalaga** [comment]
> Thus I propose the following ZIP structure
> 
> ```
> mvnd-0.0.2-darwin-amd64/
> mvnd-0.0.2-darwin-amd64/bin
> mvnd-0.0.2-darwin-amd64/bin/mvnd
> mvnd-0.0.2-darwin-amd64/daemon/
> mvnd-0.0.2-darwin-amd64/daemon/boot/
> mvnd-0.0.2-darwin-amd64/daemon/lib/
> mvnd-0.0.2-darwin-amd64/daemon/bin
> ...
> ```

Sounds good!

**Comment by ppalaga** [comment]
The bottom line is that we already have `mvnd` shell script in `bin`: https://github.com/mvndaemon/mvnd/tree/master/daemon/src/main/distro/bin 

I think we could rename it to `mvnd.sh` for now and keep distributing it as a fallback option.

File: client/src/assembly/mvnd-zip.xml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    <id>bin</id>
+    <formats>
+        <format>zip</format>
+    </formats>
+    <fileSets>
+        <fileSet>
+            <directory>${project.build.directory}</directory>
+            <outputDirectory>/</outputDirectory>
+            <includes>
+                <include>mvnd</include>
```
[the remark this answers is not in the document]
Comment (helpermethod) [reply client/src/assembly/mvnd-zip.xml:30]: > The bottom line is that we already have `mvnd` shell script in `bin`: https://github.com/mvndaemon/mvnd/tree/master/daemon/src/main/distro/bin
> 
> I think we could rename it to `mvnd.sh` for now and keep distributing it as a fallback option.

So basically the `mvnd.sh` script should be moved the `bin` folder next to the `mvnd` binary?

[the remark this answers is not in the document]
Comment (ppalaga) [reply client/src/assembly/mvnd-zip.xml:30]: Yes

**Comment by helpermethod** [comment]
Hi @ppalaga, haven't heard from the main SDKMAN! maintainer yet. Publishing the distribution and installing client and server separately seems doable but I would like to discuss this first in the SDKMAN! cli-development slack channel.

Maybe as a first step we could distribute only the native client, and in the next iteration the full distribution?

**Comment by ppalaga** [comment]
> Hi @ppalaga, haven't heard from the main SDKMAN! maintainer yet. Publishing the distribution and installing client and server separately seems doable but I would like to discuss this first in the SDKMAN! cli-development slack channel.

Not 100% sure what you mean by "installing client and server separately". My vision is to have one zip package for each platform that will contain both client and server. The client binary will be in the `bin` folder of the zip. In that way, our ZIPs will not be much different from the stock Maven ZIP. That would have the advantage, that the installation via sdkman can work the same way as with stock Maven. I hope I am not making any false assumptions here.

Clearly, we will have to adjust the way how the client finds the server directory. Currently, the client checks the `MVND_HOME` env var, `mvnd.home` system property and `~/.m2/mvnd.properties` respectively. For the "full distro unzip installation" to work flawlessly, we'll have to add some relative lookup starting from the location of `mvnd` executable itself.

> Maybe as a first step we could distribute only the native client, and in the next iteration the full distribution?

I do not think the temporary solution is worth the trouble. I like the idea of the full distro much better. 

It looks like I could find some time for this tomorrow. So I can take over preparing the distro, unless you want to finish it yourself?

**Comment by helpermethod** [comment]
So, I faced multiple problems:

* with the `provisio-maven-plugin` I couldn't find out how to reference the native image build in the `client` project, as this is no regular artifact
* with the `maven-assembly-plugin` this is easy via a `<moduleSet>` but all the other stuff is much trickier; e.g. you need to extract the `.tar.gz` dependency using the `maven-dependency-plugin` in order to strip of the root directory

So I'm giving up for today. Let's hope you have more luck :).

**Comment by helpermethod** [comment]
I've committed my last almost working revision. The only thing not working correctly is the unpacking of the tar gz archive. It's always extracting the root directory, which needs to stripped from the files' path.

**Comment by ppalaga** [comment]
Ups, looks like we worked in parallel. I created https://github.com/mvndaemon/mvnd/pull/55

**Comment by helpermethod** [comment]
Your solution looks better than mine, you can just delete my branch :).

**Comment by ppalaga** [comment]
Could you please have a look that the zips built in the other PR are as expected? https://github.com/mvndaemon/mvnd/pull/55/checks 
image The caveat is that they require Java 11 at runtime, which I'd like to address now.

**Pull request closed without merging**